### PR TITLE
:ghost: Refactor tags table to remove deprecated legacy table dep

### PR DESCRIPTION
**Pull request by ibolton336**
Working towards resolving:
https://github.com/konveyor/tackle2-ui/issues/1283

### CODE REVIEW

**Review by rszwajko — COMMENTED** [review]
1. filtering by tag color didn't work for me
2. color column has incorrect header (see below)
image

File: client/src/app/pages/controls/tags/components/tag-table.tsx [hunk 1]
```diff
@@ -94,9 +81,7 @@ export const TagTable: React.FC<TabTableProps> = ({
           const rowActions = defaultActions(row);
           return (
             <Tr>
-              {row.cells?.map((cell: any) => (
-                <Td>{cell.title}</Td>
-              ))}
+              {row.cells?.map((cell: any) => <Td>{cell.title}</Td>)}
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/components/tag-table.tsx:84]: it seems we can use `tag.name` directly without wrapping in `IRow` and `ICell`

File: client/src/app/pages/controls/tags/tags.tsx [hunk 2]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+    expandableVariant: "single",
+    filterCategories: [
+      {
+        categoryKey: "tags",
+        title: t("terms.name"),
+        type: FilterType.multiselect,
+        placeholderText:
+          t("actions.filterBy", {
+            what: t("terms.name").toLowerCase(),
+          }) + "...",
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:183]: 1. the triple dots (`"..."`) seems not needed at standard screen resolution
2. any string concatenations may clash with translation

File: client/src/app/pages/controls/tags/tags.tsx [hunk 3]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+          return tagNames || "";
         },
-        {
-          title: <Color hex={categoryColor} />,
-        },
-        {
-          title: item.tags ? item.tags.length : 0,
+        selectOptions: dedupeFunction(
+          tagCategories
+            ?.map((tagCategory) => tagCategory?.tags)
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:194]: 1. if we suspect that the array is nullish then we should continue with the `?.` operator till the end
2. how about:  ` tagCategory?.tags ?? []`
3. `flatMap`?

File: client/src/app/pages/controls/tags/tags.tsx [hunk 4]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+            .flat()
+            .filter((tag) => tag && tag.name)
+            .map((tag) => ({ key: tag?.name, value: tag?.name }))
+            .concat(
+              tagCategories?.map((tagCategory) => ({
+                key: tagCategory?.name,
+                value: tagCategory?.name,
+              }))
+            )
+            .sort((a, b) => {
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:204]: we can use `localeNumericCompare`  from the utils

File: client/src/app/pages/controls/tags/tags.tsx [hunk 5]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+      paginationProps,
+      tableProps,
+      getThProps,
+      getTrProps,
+      getTdProps,
+    },
+    expansionDerivedState: { isCellExpanded },
+  } = tableControls;
 
+  const categoryColor =
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:267]: unused constant

File: client/src/app/pages/controls/tags/tags.tsx [hunk 6]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+                      onClick={() => setTagModalState("create")}
+                    >
+                      {t("actions.createTag")}
+                    </Button>
+                  </ToolbarItem>
+                  <ToolbarItem>
+                    <Button
+                      type="button"
+                      id="create-tag-category"
+                      aria-label="Create tag category"
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:304]: the label is not translated

File: client/src/app/pages/controls/tags/tags.tsx [hunk 7]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+              <ToolbarItem {...paginationToolbarItemProps}>
+                <SimplePagination
+                  idPrefix="tag-category-table"
+                  isTop
+                  paginationProps={paginationProps}
+                />
+              </ToolbarItem>
+            </ToolbarContent>
+          </Toolbar>
+          <Table {...tableProps} aria-label="Tag category table">
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:322]: the label is not translated

File: client/src/app/pages/controls/tags/tags.tsx [hunk 8]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+                        <Td
+                          width={10}
+                          {...getTdProps({ columnKey: "tagCount" })}
+                        >
+                          {tagCategory.tags?.length || 0}
+                        </Td>
+                        <Td width={20}>
+                          <AppTableActionButtons
+                            isDeleteEnabled={!!tagCategory.tags?.length}
+                            tooltipMessage="Cannot delete non empty tag category"
```
Comment (rszwajko) [inline client/src/app/pages/controls/tags/tags.tsx:389]: the label is not translated

File: client/src/app/pages/controls/tags/tags.tsx (hunk 6 above)
Comment (sjd78) [reply client/src/app/pages/controls/tags/tags.tsx:304]: To be snarky and pedantic -- it's not externalized

File: client/src/app/pages/controls/tags/tags.tsx [hunk 9]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+                        rowIndex={rowIndex}
+                      >
+                        <Td width={25} {...getTdProps({ columnKey: "name" })}>
+                          {tagCategory.name}
+                        </Td>
+                        <Td width={10} {...getTdProps({ columnKey: "rank" })}>
+                          {tagCategory.rank}
+                        </Td>
+                        <Td width={10} {...getTdProps({ columnKey: "color" })}>
+                          <Color hex={categoryColor} />
```
Comment (sjd78) [inline client/src/app/pages/controls/tags/tags.tsx:374]: Something odd with this component such that it doesn't line up with the other text in the row.

**Review by sjd78 — COMMENTED** [review]
I don't think any of my comments are blockers, but I'm gonna have another look shortly.

File: client/src/app/pages/controls/tags/components/tag-table.tsx [hunk 10]
```diff
@@ -8,101 +8,53 @@ import {
...
       <Tbody>
-        {rows.map((row: IRow) => {
-          const rowActions = defaultActions(row);
-          return (
-            <Tr>
-              {row.cells?.map((cell: any) => (
-                <Td>{cell.title}</Td>
-              ))}
+        {(tagCategory.tags || [])
+          .sort((a, b) => a.name.localeCompare(b.name))
```
Comment (sjd78) [inline client/src/app/pages/controls/tags/components/tag-table.tsx:38]: Another place to use `localeNumericCompare()`

FWIW, `localeNumbericCompare()` could be changed a bit to have the locale be defaulted to "i18n.language" and ignore that part here...(could also be another PR)
```jsx
/**
 * Uses native string localCompare method with numeric option enabled.
 *
 * @param locale to be used by string compareFn
 */
export const localeNumericCompare = (
  a: string,
  b: string,
  locale: string = i18n.language
): number => a.localeCompare(b, locale, { numeric: true });
```

**Review by sjd78 — COMMENTED** [review]
One more sorting change and good to go from me.

File: client/src/app/pages/controls/tags/components/tag-table.tsx (hunk 10 above)
Comment (rszwajko) [reply client/src/app/pages/controls/tags/components/tag-table.tsx:38]: I vote for a small nice PR!

**Review by rszwajko — APPROVED** [review]
Looks OK to me!

File: client/src/app/pages/controls/tags/tags.tsx (hunk 9 above)
Comment (sjd78) [reply client/src/app/pages/controls/tags/tags.tsx:374]: but also not a regression -- so that can be a followup PR as well

File: client/src/app/pages/controls/tags/components/tag-table.tsx (hunk 10 above)
Comment (ibolton336) [reply client/src/app/pages/controls/tags/components/tag-table.tsx:38]: https://github.com/konveyor/tackle2-ui/issues/1824

File: client/src/app/pages/controls/tags/tags.tsx (hunk 9 above)
Comment (ibolton336) [reply client/src/app/pages/controls/tags/tags.tsx:374]: https://github.com/konveyor/tackle2-ui/issues/1825

File: client/src/app/pages/controls/tags/tags.tsx [hunk 11]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+            .concat(tagCategoryNames)
+            .join("");
+          return tagNames || "";
         },
-        {
-          title: <Color hex={categoryColor} />,
-        },
-        {
-          title: item.tags ? item.tags.length : 0,
+        selectOptions: dedupeFunction(
```
[the remark this answers is not in the document]
Comment (ibolton336) [reply client/src/app/pages/controls/tags/tags.tsx:196]: https://github.com/konveyor/tackle2-ui/issues/1826

File: client/src/app/pages/controls/tags/tags.tsx [hunk 12]
```diff
@@ -157,281 +152,283 @@ export const Tags: React.FC = () => {
...
+        placeholderText:
+          t("actions.filterBy", {
+            what: t("terms.name").toLowerCase(),
+          }) + "...",
+        getItemValue: (item) => {
+          const tagCategoryNames = item.name?.toString() || "";
+          const tagNames = item?.tags
+            ?.map((tag) => tag.name)
+            .concat(tagCategoryNames)
+            .join("");
```
[the remark this answers is not in the document]
Comment (ibolton336) [reply client/src/app/pages/controls/tags/tags.tsx:193]: https://github.com/konveyor/tackle2-ui/issues/1827
